src/deploy/functions/params.ts

```diff
@@ -0,0 +1,332 @@
+ * - the value returned by interactively prompting the user
+ *   - the default value of the prompt comes from the SDK via param.default, which may be a literal value or a CEL expression
+ *   - if the default CEL expression is not resolvable--it depends on a param whose value is not yet known--we throw an error
+ *   - yes, this means that the same set of params may or may not throw depending on the order the SDK provides them to us in
+ *   - after prompting, the resolved value of the param is written to the most specific .env file available
+ */
+export async function resolveParams(
+  params: Param[],
+  projectId: string,
+  userEnvs: Record<string, ParamValue>
+): Promise<Record<string, ParamValue>> {
+  const paramValues: Record<string, ParamValue> = {};
+
+  for (const param of params.filter((param) => userEnvs.hasOwnProperty(param.param))) {
+    if (!canSatisfyParam(param, userEnvs[param.param])) {
+      throw new FirebaseError(
+        "Parameter " +
+          param.param +
+          " resolved to value from dotenv files " +
+          userEnvs[param.param] +
+          " of wrong type"
+      );
+    }
+    paramValues[param.param] = userEnvs[param.param];
+  }
+
+  for (const param of params.filter((param) => !userEnvs.hasOwnProperty(param.param))) {
+    let paramDefault: ParamValue | undefined = param.default;
+    if (paramDefault && isCEL(paramDefault)) {
+      paramDefault = resolveDefaultCEL(param.type, paramDefault, paramValues);
+    }
+    if (paramDefault && !canSatisfyParam(param, paramDefault)) {
+      throw new FirebaseError(
+        "Parameter " + param.param + " has default value " + paramDefault + " of wrong type"
+      );
+    }
+    paramValues[param.param] = await promptParam(param, paramDefault);
+
+    // TODO(vsfan@): Once we have writeUserEnvs in functions/env.ts implemented, call it to persist user-provided params
+  }
+
+  return paramValues;
+}
+
+/**
+ * Returns the resolved value of a user-defined Functions parameter.
+ * Functions params are defined by the output of the Functions SDK, but their value is not set until deploy-time.
+ *
+ * For most param types, we check the contents of the dotenv files first for a matching key, then interactively prompt the user.
+ * When the CLI is running in non-interactive mode or with the --force argument, it is an error for a param to be undefined in dotenvs.
+ */
+async function promptParam(param: Param, resolvedDefault?: ParamValue): Promise<ParamValue> {
+  if (param.type === "string") {
+    return promptStringParam(param, resolvedDefault as string | undefined);
+  } else if (param.type === "int") {
+    return promptIntParam(param, resolvedDefault as number | undefined);
+  }
+  assertExhaustive(param);
+}
+
+async function promptStringParam(param: StringParam, resolvedDefault?: string): Promise<string> {
+  if (!param.input) {
+    const defaultToText: TextInput<string> = { type: "text", text: {} };
+    param.input = defaultToText;
```

Comment:
fyi: the reason you need to cast here is because `text` is a string, not a literal. Other techniques you can do:

1. Assign to `param.input` directly while keeping `"text"` a literal string
2. Keep `"text"` from type-erasing to string by saying `{ type: "text", text: {}} as const`
3. Keep as is

Just an FYI, not a request to change.